[SE-0279][Parser] Add support for multiple trailing closures syntax

xedin opened the pull request:
Accept trailing closures in following form:

```swift
foo {
  <label-1>: { ... }
  <label-2>: { ... }
  ...
  <label-N>: { ... }
}
```

Consider each labeled block to be a regular argument to a call or subscript,
so the result of parser looks like this:

```swift
foo(<label-1>: { ... }, ..., <label-N>: { ... })
```

Note that in this example parens surrounding parameter list are implicit
and for the cases when they are given by the user e.g.

```swift
foo(bar) {
  <label-1>: { ... }
  ...
}
```

location of `)` is changed to a location of `}` to make sure that call
"covers" all of the transformed arguments and parser result would look
like this:

```swift
foo(bar,
  <label-1>: { ... }
)
```

Resolves: rdar://problem/59203764

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] Extend TupleExpr to support multiple trailing closures

xedin committed:
[AST] Remove unused "trailing closure" argument from `totalSizeToAlloc` and `initializeCallArguments`

xedin committed:
[AST] Add support for multiple trailing closures to the parser/expressions

xedin committed:
[Parser] Adjust `parseExprList` to return multiple trailing closures

Also extend returned object from simplify being an expression to
`TrailingClosure` which has a label, label's source location and
associated closure expression.

xedin committed:
[Parse] Account that there could be multiple trailing closures which parsing

xedin committed:
[Parser] Add support for multiple trailing closures syntax

Accept trailing closures in following form:

```swift
foo {
  <label-1>: { ... }
  <label-2>: { ... }
  ...
  <label-N>: { ... }
}
```

Consider each labeled block to be a regular argument to a call or subscript,
so the result of parser looks like this:

```swift
foo(<label-1>: { ... }, ..., <label-N>: { ... })
```

Note that in this example parens surrounding parameter list are implicit
and for the cases when they are given by the user e.g.

```swift
foo(bar) {
  <label-1>: { ... }
  ...
}
```

location of `)` is changed to a location of `}` to make sure that call
"covers" all of the transformed arguments and parser result would look
like this:

```swift
foo(bar,
   <label-1>: { ... }
)
```

Resolves: rdar://problem/59203764

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3127,19 +3143,88 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
+
+  auto text = P.Tok.getText();
+  if (text[0] == '$')
+    return false;
```
You could this before starting the backtracking.
```cpp
if (!P.peekToken().canBeArgumentLabel() || P.peekToken().is(tok::kw__) ||
    P.peekToken().getText()[0] == '$')
  return false;

Parser::BacktrackingScope backtrack(P);
P.consumeToken(tok::l_brace);
P.consumeToken();
```

xedin replied:
I can’t do that because its labeled blocks enclosed in braces, so I need to consume or peek through the left brace first somehow first...

xedin replied:
Actually I take that back, I could do that to avoid backtracking

rintaro replied:
But `peekToken()` can look through the brace..

xedin replied:
Fixed!

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3127,19 +3143,88 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
+    SourceLoc &LBrace, SourceLoc &RBrace,
+    SmallVectorImpl<TrailingClosure> &closures) {
+  // Consume '{' of the trailing closure
+  LBrace = consumeToken();
```
`consumeToken(tok::l_brace)`, just to make sure.

xedin replied:
Ah, thanks!

xedin replied:
Fixed! :)

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3127,19 +3143,88 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
+  // There could N labeled closures depending on the number of arguments.
+  do {
+    if (!(Tok.canBeArgumentLabel() && peekToken().is(tok::colon)))
+      return makeParserError();
```
Could you keep the already-parsed-closures in the AST somehow? Otherwise, many SourceKit functionalities will be messed up because it cannot know the structure of the source code.

rintaro replied:
Also, when you detect errors in the braces, you should skip to `}` or next stmt (using `skipUntilDeclStmtRBrace()`), and hopefully parse `}` as the `RBrace`.

xedin replied:
This method is used to peek ahead and check whether what follows represents multiple closures, if not it’s just going to parse it as a regular closure, it doesn’t affect the state of parser or drop already parsed nodes...

xedin replied:
I think such preservation already happens because closures array with be still returned but with error status. I am going to add skipping though

rintaro replied:
For example:
```
doSomething() {
  labelA: { ... }
  1 + 2
}
```
In this case, `isBlockOfMultipleTrailingClosures()` returns true, right?. We want to skip `1 + 2`, and keep other parts in the AST.

xedin replied:
Sure, it does make sense to skip all of the invalid bits!

xedin replied:
Fixed!

xedin committed:
[Parser] NFC: Move multiple trailing closures tests into a separate file

xedin committed:
[Parse] Introduce tailored diagnostics for an invalid trailing closures block

If block is not just a list of labeled closures produce a tailored
diagnostic and skip over invalid code.

xedin committed:
[Parser] Check for first label in trailing closure block without backtracking

rintaro committed:
[Syntax] Add declaration for multiple trailing closure

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3122,19 +3138,95 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
+  // If closure doesn't start from a label there is no reason
+  // to look any farther.
+  {
+    auto nextTok = P.peekToken();
```
`const auto &nextTok = `
we don't need to copy it.

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3122,19 +3138,95 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
+  if (!P.Tok.is(tok::colon))
+    return false;
+
+  P.consumeToken(); // Consume `:`
```
Can be 
```cpp
// Consume ':'.
if (!P.consumeIf(tok::colon))
  return false;
```

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3122,19 +3138,95 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
+  parseMatchingToken(tok::r_brace, RBrace,
+                     diag::expected_multiple_closures_block_rbrace, LBrace);
+
+  return makeParserSuccess();
```
Could you propagate the status from `parseExprClosure()`s? Like:
```cpp
ParserStatus Status;
do {
   ...
   auto closure = parseExprClosure();
   Status |= closure;
   ...
} while (...);

if (parseMatchingToken(...)) {
  Status.setIsParseError()
}

return Status;
```
Because we want to know if a code completion happens.
You don't need to care about code completion at the labels position, because we can do it later.

xedin committed:
[Parser] Propagate status information from each closure in the trailing block

rintaro committed:
[Parse] Report "success" if '}' is found for multiple trailing closures

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3122,19 +3138,106 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
+
+    consumeToken(); // Consume ':'
+
+    auto closure = parseExprClosure();
```
`Tok.is(tok::l_brace)` check is needed before calling this. `parseExprClosure()` assert it.

rintaro replied:
Or just `parseExpr()` and diagnose if the result is not `ClosureExpr`/`CaptureListExpr`

rintaro commented on lib/AST/Expr.cpp:
```diff
@@ -997,11 +1000,13 @@ swift::packSingleArgument(ASTContext &ctx, SourceLoc lParenLoc,
+    auto &trailingClosure = trailingClosures.front();
+    auto arg =
+        new (ctx) ParenExpr(lParenLoc, trailingClosure.ClosureExpr, rParenLoc,
+                            /*hasTrailingClosure=*/true);
```
How do we distinguish `foo { 1 }` and `foo { fn: { 1 } }`? Also, this loses the parsed argument label.
```swift
func foo(bar: () -> Int) {}
foo { bang: { 12 } }
```
cannot be diagnosed.

Commit:
Multiple trailing closure fixes

- Diagnose non-closure-literals after label + colon
- Form a TupleExpr, rather than a ParenExpr for a single labelled trailing closure.

Commit:
Add indentation support for multiple trailing closures.

Resolves rdar://problem/60250267

rintaro committed:
[CodeCompletion] Completion inside multiple trailing closure

rdar://problem/59688477

xedin committed:
[Parse] Allow multiple trailing closures to be delimited by `_:`

Syntax like `foo { _: { ... } }` is going to be parsed as a single
trailing closure.
